### PR TITLE
Perf ui

### PR DESCRIPTION
**Pull request by Therzok**
If we want this in, we need to:
~1. Merge this: https://github.com/mono/gtk-sharp/pull/200 - the reflection code would be too complicated to take into account for the Objects dictionary being in the Object class, plus, ToggleRef is not publicly accessible, therefore we need to use fully untyped reflection until we can get the array of pointers.~
~2. Bump mono to a build with this gtk# version~

Not needed to do anything special anymore.

### CODE REVIEW

**Comment by Therzok** [comment]
https://www.screencast.com/t/7wpPlkpBN - Check the bottom right
https://www.screencast.com/t/5qW4FJDng3

File: main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/LeakHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+		static int remainingEqualChangedCount = 5;
+		static bool SummaryTimeoutHandler ()
+		{
+			int gobjCount = GObjectDict.Count;
+			int nsobjCount = NSObjectDict.Count;
+
+			GC.Collect ();
+			GC.Collect ();
+			GC.Collect ();
+			GC.WaitForPendingFinalizers ();
```
Comment (alanmcgovern) [inline main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/LeakHelpers.cs]: The finalizers for Gtk can only do their job properly if the main loop is pumping. We might have better accuracy if you do 3 GC calls and wait for finaliser calls, separated by 1 second, and then perform the analysis of live objects

File: main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/Options.cs [hunk 2]
```diff
@@ -31,6 +31,10 @@ namespace PerformanceDiagnosticsAddIn
 	public static class Options
 	{
 		public static readonly ConfigurationProperty<string> OutputPath = ConfigurationProperty.Create ("PerformanceDiagnosticsAddIn.OutputPath", System.IO.Path.Combine (Environment.GetFolderPath (Environment.SpecialFolder.Personal), "Desktop"));
+		//public static readonly ConfigurationProperty<bool> ShowUICounters = ConfigurationProperty.Create ("PerformanceDiagnosticsAddIn.ShowUICounters", false);
+
+		public static readonly bool HasMemoryLeakFeature = System.IO.File.Exists (System.IO.Path.Combine (Environment.GetFolderPath (Environment.SpecialFolder.Personal), ".xs-enable-leak-check"));
+		public static readonly bool HasMemoryLeakFeaturePad = HasMemoryLeakFeature && System.IO.File.Exists (System.IO.Path.Combine (Environment.GetFolderPath (Environment.SpecialFolder.Personal), ".xs-enable-leak-check-pad"));
```
Comment (alanmcgovern) [inline main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/Options.cs]: By convention we don't prefix with . as it's hard to create on windows

File: main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/LeakHelpers.cs (hunk 1 above)
Comment (Therzok) [inline main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/LeakHelpers.cs]: This method is queued recursively, guaranteeing that for example, unrefs done on chained objects (Gtk.TextTagTable + Gtk.TextTag) both get finalized.

Comment (Therzok) [inline main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/LeakHelpers.cs]: We queue the timeout again with `return true` to run a second time.

File: main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/Options.cs (hunk 2 above)
Comment (Therzok) [inline main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/Options.cs]: Okay, sure

File: main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/LeakHelpers.cs (hunk 1 above)
Comment (alanmcgovern) [inline main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/LeakHelpers.cs]: Oh i see now :P It helps when i have a large screen to read the entire diff!

File: main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/Options.cs [hunk 3]
```diff
@@ -31,6 +31,10 @@ namespace PerformanceDiagnosticsAddIn
 	public static class Options
 	{
 		public static readonly ConfigurationProperty<string> OutputPath = ConfigurationProperty.Create ("PerformanceDiagnosticsAddIn.OutputPath", System.IO.Path.Combine (Environment.GetFolderPath (Environment.SpecialFolder.Personal), "Desktop"));
+		//public static readonly ConfigurationProperty<bool> ShowUICounters = ConfigurationProperty.Create ("PerformanceDiagnosticsAddIn.ShowUICounters", false);
```
Comment (alanmcgovern) [inline main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/Options.cs]: Why is this new code commented out?

File: main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/GlobalOptionsDialog.cs [hunk 4]
```diff
@@ -29,13 +30,24 @@ public override Control CreatePanelWidget()
...
 			vbox.ShowAll();
 			return vbox;
 		}
 
 		public override void ApplyChanges()
 		{
 			Options.OutputPath.Value = folderEntry.Path;
+
+			//if (Options.HasMemoryLeakFeature)
+				//Options.ShowUICounters.Value = uiCountersCheck.Active;
```
Comment (alanmcgovern) [inline main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/GlobalOptionsDialog.cs]: new code commented out?

Comment (Therzok) [inline main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/GlobalOptionsDialog.cs]: Was the intention of having something for users, not just us.

**Comment by alanmcgovern** [comment]
👍 then!

File: main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/PerformanceDiagnosticsAddIn.csproj [hunk 5]
```diff
@@ -10,7 +10,7 @@
     <OutputType>Library</OutputType>
     <RootNamespace>PerformanceDiagnosticsAddIn</RootNamespace>
     <AssemblyName>PerformanceDiagnosticsAddIn</AssemblyName>
-    <TargetFrameworkVersion>v4.6.1</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.7</TargetFrameworkVersion>
```
Comment (Therzok) [inline main/src/addins/PerformanceDiagnostics/PerformanceDiagnostics/PerformanceDiagnosticsAddIn.csproj]: Note to self: Revert this to 4.6.1 and fixup ValueTuple package.

**Pull request merged**